daemon/tests/harness/flow.rs

```diff
@@ -62,15 +37,53 @@ pub async fn next<T>(rx: &mut watch::Receiver<T>) -> Result<T>
 where
     T: Clone,
 {
-    rx.changed()
+    next_with(rx, |v| Some(v)).await
+}
+
+/// Returns watch channel value upon change
+pub async fn next_with<T, U>(
+    rx: &mut watch::Receiver<T>,
+    filter_map: impl Fn(T) -> Option<U>,
+) -> Result<U>
+where
+    T: Clone,
+{
+    let wait_until_predicate = async {
+        loop {
+            rx.changed().await?;
+
+            let current = rx.borrow().clone();
+
+            if let Some(val) = filter_map(current) {
+                return anyhow::Ok(val);
+            }
+        }
+    };
+
+    let val = wait_until_predicate
         .timeout(NEXT_WAIT_TIME)
         .await
         .with_context(|| {
             format!(
-                "No change in channel within {} seconds",
+                "Value channel did not satisfy predicate within {} seconds",
                 NEXT_WAIT_TIME.as_secs()
             )
         })??;
 
-    Ok(rx.borrow().clone())
+    Ok(val)
+}
+
+/// Drop-in filter-map function for [`next_with`] to check the state of the CFD in a list of CFDs.
+///
+/// # Panics
+///
+/// If there is more than one CFD in the list. This is unsupported and unexpected by our test
+/// framework.
+pub fn one_cfd_with_state(expected_state: CfdState) -> impl Fn(Vec<Cfd>) -> Option<Cfd> {
+    move |cfds: Vec<Cfd>| match cfds.as_slice() {
+        [one] if one.state == expected_state => Some(one.clone()),
+        [_one_that_doesnt_match_state] => None,
+        [] => None,
+        _more_than_one => panic!("More than one CFD in feed!"),
```

Comment:
never seen this syntax before - what does it mean? are the `[]` destructuring the array of cfds?

Reply:
This is called sub-slice pattern matching: https://github.com/rust-lang/rfcs/blob/master/text/2359-subslice-pattern-syntax.md